fendermint/app/src/snapshot.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::BlockHeight;
+use fendermint_vm_interpreter::fvm::state::{FvmStateParams, Snapshot};
+use fvm_ipld_blockstore::Blockstore;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+/// The config params for snapshot
+#[derive(Clone)]
+pub struct SnapshotConfig {
+    path: String,
+    period: BlockHeight,
+}
+
+/// Manager the snapshot of the entire state tree and state params at certain blocks.
+#[derive(Clone)]
+pub struct SnapshotManager {
+    config: SnapshotConfig,
+    lock: Arc<AtomicBool>,
+}
+
+const STARTED: bool = true;
+
+impl SnapshotManager {
+    pub fn new(config: SnapshotConfig) -> Self {
+        Self {
+            config,
+            lock: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    pub fn should_run(&self, height: BlockHeight) -> bool {
+        height % self.config.period == 0
+    }
+
+    pub fn start<BS: Blockstore + 'static + Send>(
+        &self,
+        block_height: BlockHeight,
+        state_params: FvmStateParams,
+        store: BS,
+    ) -> bool {
+        tracing::debug!("attempting to trigger snapshot at height: {block_height}");
+
+        let snapshot = match Snapshot::new(store, state_params, block_height) {
+            Ok(s) => s,
+            Err(e) => {
+                tracing::warn!("cannot start snapshot at block: {block_height} due to: {e}");
+                return false;
+            }
+        };
+
+        match self
+            .lock
+            .compare_exchange(!STARTED, STARTED, Ordering::SeqCst, Ordering::SeqCst)
+        {
+            Ok(_) => {}
+            Err(_) => {
+                tracing::info!(
+                    "triggering snapshot at height: {block_height} fails as it's already running"
+                );
+                return false;
+            }
+        };
+
+        let path = self.config.path.clone();
```

Comment:
This `path` seems like it's going to be the same for every snapshot. I see from the [implementation](https://github.com/consensus-shipyard/fendermint/blob/5faf85d841b94d79cd6b936ab922c1005525b2df/fendermint/vm/interpreter/src/fvm/state/snapshot.rs#L82-L83) that it is a path a file, not a directory, which means that every snapshot will overwrite the previous one. 

The problem is that this happens as soon as the `write_car` method starts, which has multiple negative consequences:
* If anything goes wrong, we lost our only working snapshot
* The file is created and it looks like a ready snapshot even though it's still being written to
* If anything goes wrong we might end up with a partial, invalid snapshot
* It's impossible to tell from the file name what the snapshot is for

Ideally the path would point at a directory, where we can store snapshots that bear the name of the epoch for which they were made. When we start writing a snapshot, it should go into a temporary file, which is renamed to its final form after it's ready. It should be easy to filter and sort snapshots in the ABCI methods to only offer the latest complete ones, and to remove the oldest ones when a new one is ready. We should also be able to store checksum files next to the snapshots in the directory.